File: Week_02/589/n_ary_tree_preorder_traversal2.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []

        stack = [root]
        while stack:
            node = stack.pop()
            if node is not None:
                res.append(node.val)
                for i in range(len(node.children) - 1, -1, -1):
                    stack.append(node.children[i])
        return res

File: Week_02/589/n_ary_tree_preorder_traversal2.go
package main

func preorder2(root *Node) []int {
	res := make([]int, 0)
	stack := []*Node{root}

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if node != nil {
			res = append(res, node.Val)
			for i := len(node.Children) - 1; i >= 0; i-- {
				stack = append(stack, node.Children[i])
			}
		}
	}
	return res
}
